extending/plugins.rst

```diff
@@ -0,0 +1,134 @@
+.. _plugins:
+
+Plugins [EXPERIMENTAL]
+======================
+
+The Conan plugins is a feature intended to extend the Conan functionalities and let users customize the client behavior at determined
+points.
+
+Plugin structure
+----------------
+
+Plugins are Python files containing **pre** and **post** functions that will be executed prior and after a determined task performed by the
+Conan client. Those tasks could be Conan commands, recipe interactions such as exporting or packaging or interactions with the remotes.
+
+Here you can see an example of a simple plugin:
+
+.. code-block:: python
+   :caption: *example_plugin.py*
+
+    from conans import tools
+
+
+    def pre_export(output, conanfile, conanfile_path, reference, **kwargs):
+        conanfile_content = tools.load(conanfile_path)
+        test = "[RECIPE METADATA]"
+        metadata_error = False
+        for field in ["url", "license", "description"]:
+            field_value = getattr(conanfile, field, None)
+            if not field_value:
+                metadata_error = True
+                output.error("%s Conanfile doesn't have '%s'. It is recommended to add it as attribute"
+                            % (test, field))
+
+    def pre_source(output, conanfile, conanfile_path, **kwargs):
+        conanfile_content = tools.load(conanfile_path)
+        if "def source(self):" in conanfile_content:
+            test = "[IMMUTABLE SOURCES]"
+            valid_content = [".zip", ".tar", ".tgz", ".tbz2", ".txz"]
+            invalid_content = ["git checkout master", "git checkout devel", "git chekcout develop"]
+            if "git clone" in conanfile_content and "git checkout" in conanfile_content:
+                fixed_sources = True
+                for invalid in invalid_content:
+                    if invalid in conanfile_content:
+                        fixed_sources = False
+            else:
+                fixed_sources = False
+                for valid in valid_content:
+                    if valid in conanfile_content:
+                        fixed_sources = True
+
+            if not fixed_sources:
+                output.error("%s Source files does not come from and immutable place. Checkout to a "
+                            "commit/tag or download a compressed source file" % test)
```

Comment:
I find this `pre_source` example quite difficult. I like the intention, but I think it is very tricky

Reply:
Agree, would be better to keep the example more simple and obvious to read.